lib/evmone/eof.cpp

```diff
@@ -90,37 +151,270 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+    }
+
+    return types;
+}
+
 EOFValidationError validate_instructions(evmc_revision rev, bytes_view code) noexcept
 {
     assert(!code.empty());  // guaranteed by EOF headers validation
 
-    size_t i = 0;
-    uint8_t op = code[0];
-    while (i < code.size())
+    const auto& cost_table = baseline::get_baseline_cost_table(rev, 1);
+
+    for (size_t i = 0; i < code.size(); ++i)
     {
-        op = code[i];
-        const auto& since = instr::traits[op].since;
-        if (!since.has_value() || *since > rev)
+        const auto op = code[i];
+        if (cost_table[op] == instr::undefined)
             return EOFValidationError::undefined_instruction;
 
-        i += instr::traits[op].immediate_size;
-        ++i;
-    }
+        if (op == OP_RJUMPV)
+        {
+            if (i + 1 >= code.size())
+                return EOFValidationError::truncated_instruction;
 
-    if (!instr::traits[op].is_terminating)
-        return EOFValidationError::missing_terminating_instruction;
+            const auto count = code[i + 1];
+            if (count < 1)
+                return EOFValidationError::invalid_rjumpv_count;
+            i += static_cast<size_t>(1 /* count */ + count * 2 /* tbl */);
+        }
+        else
+            i += instr::traits[op].immediate_size;
+
+        if (i >= code.size())
+            return EOFValidationError::truncated_instruction;
+    }
 
     return EOFValidationError::success;
 }
 
+/// Validates that that we don't rjump inside an instruction's immediate.
+/// Requires that the input is validated against truncation.
+bool validate_rjump_destinations(bytes_view code) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    const auto code_size = code.size();
+    // list of all possible absolute rjumps destinations positions
+    std::vector<size_t> rjumpdests;
+    // bool map of immediate arguments positions in the code
+    std::vector<bool> immediate_map(code_size);
+
+    const auto check_rjump_destination = [code_size](auto rel_offset_data_it,
+                                             size_t post_pos) -> std::optional<size_t> {
+        const auto rel_offset = read_int16_be(rel_offset_data_it);
+        const auto jumpdest = static_cast<int32_t>(post_pos) + rel_offset;
+        if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+            return {};
+        else
+            return static_cast<size_t>(jumpdest);
+    };
+
+    for (size_t i = 0; i < code_size; ++i)
+    {
+        const auto op = code[i];
+        size_t imm_size = instr::traits[op].immediate_size;
+
+        if (op == OP_RJUMP || op == OP_RJUMPI)
+        {
+            if (auto jumpdest_opt = check_rjump_destination(&code[i + 1], i + 3); jumpdest_opt)
+                rjumpdests.push_back(jumpdest_opt.value());
+            else
+                return false;
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
```

Comment:
This constant could be useful for RJUMP and RJUMPI above too:
```
            imm_size = REL_OFFSET_SIZE;
            const size_t post_pos = i + 1 + imm_size;
```

Reply:
xxx

Reply:
Done.